### PR TITLE
Added a dialog to prepare data for cluster analysis

### PR DESCRIPTION
**Pull request by Vitalis95**
@rdstern @shadrackkibet, this is prepare dialog for cluster analysis, issue #6195. I have added two functions; scale() and na.omit(),
Please have a look at it.

### CODE REVIEW

**Comment by shadrackkibet** [comment]
@HawardKetoyoMsatsi please test the functionality.

**Comment by HawardKetoyoMsatsi** [comment]
Fine I will take a look.

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 it says there are build errors.  This may be just me, but I can't test.  Perhaps you can fix, or I could also look at it working on your machine - shared perhaps with skype?

**Comment by shadrackkibet** [comment]
@rdstern I suggest you wait for Haward to look into this first then you can test.

File: instat/dlgCluster.vb [hunk 1]
```diff
@@ -1,3 +1,192 @@
...
+    Private Sub SetBaseFunction()
+        If rdoScaleData.Checked Then
+            If ucrChkOmitMissingRows.Checked Then
+                ucrBase.clsRsyntax.SetBaseRFunction(clsStatsNAOmitFunction)
+            Else
+                ucrBase.clsRsyntax.SetBaseRFunction(clsScaleFunction)
+            End If
+        Else
+
+        End If
```
Comment (HawardKetoyoMsatsi) [inline instat/dlgCluster.vb:143]: Make this condition complete.

File: instat/dlgCluster.vb [hunk 2]
```diff
@@ -1,3 +1,192 @@
...
+        If rdoScaleData.Checked Then
+            ucrSaveNewDataFrame.SetPrefix("scale")
+        Else
+            ucrSaveNewDataFrame.SetPrefix("distance")
+        End If
+    End Sub
+
+    Private Sub ucrCore_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverPrepareData.ControlContentsChanged, ucrSelectorPrepareData.ControlContentsChanged, ucrPnlSelectData.ControlContentsChanged
+        TestOkEnabled()
+    End Sub
```
Comment (HawardKetoyoMsatsi) [inline instat/dlgCluster.vb:178]: This sub should also handle control contents change for rdoSelectedColumn.

**Review by HawardKetoyoMsatsi — CHANGES_REQUESTED** [review]
1. I see frmMain changes in your PR kindly conform with @shadrackkibet whether you can commit frmMain main in this PR.
2. I see a problem with your frmMain resource file which is causing the build errors please take a look.

File: instat/dlgCluster.vb (hunk 2 above)
Comment (Vitalis95) [reply instat/dlgCluster.vb:178]: I will add the changes

Comment (Vitalis95) [reply instat/dlgCluster.vb:178]: @HawardKetoyoMsatsi I think the panel controls this? Radio buttons doesn't have the control content changed property.

**Comment by shadrackkibet** [comment]
@rdstern check on the functionality. 

We will sort out the changes on the main form later. Thanks

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 it says there are build errors?  Sorry @shadrackkibet I don't know how to proceed further in such a case?

**Comment by shadrackkibet** [comment]
Makes sense. Happy with your suggestions @lloyddewit

**Comment by lloyddewit** [comment]
> 
> 
> @shadrackkibet @lloyddewit does it mean that I should close this PR?

As discussed on WhatsApp:

image

**Comment by shadrackkibet** [comment]
@rdstern hold on a bit. We will sort this out soon. Thanks

**Comment by Vitalis95** [comment]
@rdstern  have a look it now, I have added the dist function. The data I am using to test is from library (cluster: voters.repub). Thanks

**Comment by shadrackkibet** [comment]
At the same time @Gioche6 could you test and review the code?

**Comment by Gioche6** [comment]
@shadrackkibet , Sure. I'll be glad to to so

**Review by rdstern — DISMISSED** [review]

**Review by lloyddewit — DISMISSED** [review]
@Vitalis95 - Thank you for the changes, I'mm happy to approve.
@shadrackkibet - I saw that you also started to review. That's good, because you have good experience with the R-Instat dialog idioms and this will help Vitalis.
We can merge after you approve. Thanks

**Comment by Vitalis95** [comment]
@shadrackkibet  have a look at it.

**Review by shadrackkibet — CHANGES_REQUESTED** [review]
Not stable yet. From a user perspective, there are various issues which we need to sort out. Like when you use a data frame with non-numeric columns you get errors. Perhaps we should try to stop this? There are also reset issues for example the radio button selected doesn't reset to the default. Should we restrict selected variables to only numerics?

**Comment by Vitalis95** [comment]
The receiver for the selected variables only allows numeric variables. I made the  default to be selection of variables as suggested by @rdstern but not the whole data frame  
The data frame should have numeric columns that is why  with non-numeric columns it gives an error. I will be later adding daisy function for non-numeric columns (factors) -  @shadrackkibet @rdstern

**Review by rdstern — APPROVED** [review]
Now seems to work.  Are there more checks or can this be merged?

**Comment by lloyddewit** [comment]
> Now seems to work. Are there more checks or can this be merged?

@shadrackkibet Thank you for reviewing this PR earlier. You had some comments. Are these now resolved and may we merge?
Thanks

**Comment by shadrackkibet** [comment]
> Not stable yet. From a user perspective, there are various issues which we need to sort out. Like when you use a data frame with non-numeric columns you get errors. Perhaps we should try to stop this? There are also reset issues for example the radio button selected doesn't reset to the default. Should we restrict selected variables to only numerics?

@Vitalis95 did anything change after this comment? Apologies if I missed out on something.

**Comment by Vitalis95** [comment]
> > Not stable yet. From a user perspective, there are various issues which we need to sort out. Like when you use a data frame with non-numeric columns you get errors. Perhaps we should try to stop this? There are also reset issues for example the radio button selected doesn't reset to the default. Should we restrict selected variables to only numerics?
> 
> @Vitalis95 did anything change after this comment? Apologies if I missed out on something.

@shadrackkibet The receiver for the selected variables only allows numeric variables. I made the default to be selected variables as suggested by @rdstern but not the whole data frame
The whole data frame allows only numeric columns that is why with non-numeric columns it gives an error, @shadrackkibet you can guide me on this , to stop the error in case of a data frame with non-numeric columns.
Thanks.

**Comment by shadrackkibet** [comment]
Hide non-numeric columns from the selector. To do this set the receiver to allow only numerics.  Pseudocode as follows `ucReceiverName.SetDataType("numeric")` also change the selector header to "Numerics" which should be something like `ucrReceivername.strSelectorHeader = "Numerics"`.

**Review by shadrackkibet — APPROVED** [review]
I approve and merge because what is here works already. Although there are still some issues, it qualifies as the initial implementation. I suggest we address them separately.

**Pull request merged**